File: python/hangman.py
import random

print("Let's play a game of hangman")
print("----------------------------")

possible_answers: list[str] = [
    "harmony",
    "venture",
    "candles",
    "journey",
    "breathe",
    "lantern",
    "whisper",
    "fortune",
    "orchard",
    "gravity",
    "twinkle",
    "beneath",
    "crimson",
    "charity",
    "freedom",
    "trouble",
    "diamond",
    "garland",
    "promise",
    "serpent",
    "miracle",
    "picture",
    "fiction",
    "harvest",
    "capture",
    "dolphin",
    "holiday",
    "passion",
    "fragile",
    "fountain",
    "eclipse",
    "twilight",
    "flicker",
    "gateway",
    "shelter",
    "forever",
]

random_answer: str = random.choice(possible_answers)

num_of_wrong_guesses_allowed: int = 9

for _ in random_answer:
    print("_", end=" ")


def print_hangman(wrong: int) -> None:
    if wrong == 0:
        print("\n")
    if wrong == 1:
        print("\n ===")
    if wrong == 2:
        print("\n    |")
        print("    |")
        print("    |")
        print("   ===")
    if wrong == 3:
        print("\n+---+")
        print("    |")
        print("    |")
        print("    |")
        print("   ===")
    elif wrong == 4:
        print("\n+---+")
        print("O   |")
        print("    |")
        print("    |")
        print("   ===")
    elif wrong == 5:
        print("\n+---+")
        print("O   |")
        print("|   |")
        print("    |")
        print("   ===")
    elif wrong == 6:
        print("\n+---+")
        print(" O  |")
        print("/|  |")
        print("    |")
        print("   ===")
    elif wrong == 7:
        print("\n+---+")
        print(" O  |")
        print("/|\\ |")
        print("    |")
        print("   ===")
    elif wrong == 8:
        print("\n+---+")
        print(" O  |")
        print("/|\\ |")
        print("/   |")
        print("   ===")
    elif wrong == 9:
        print("\n+---+")
        print(" O   |")
        print("/|\\  |")
        print("/ \\  |")
        print("    ===")


def print_word(guessed: list) -> int:
    count: int = 0
    correct_letters: int = 0
    for letter in random_answer:
        if letter in guessed:
            print(random_answer[count], end=" ")
            correct_letters += 1
        else:
            print(" ", end=" ")
        count += 1

    return correct_letters


def print_lines() -> None:
    print("\r")
    for _ in random_answer:
        print("\u203E", end=" ")


length_of_random_answer: int = len(random_answer)
num_of_wrong_guesses: int = 0
num_correct_guesses: int = 0
current_letters_guessed: list = []

while (
    num_of_wrong_guesses != num_of_wrong_guesses_allowed
    and num_correct_guesses != length_of_random_answer
):
    print("\nLetters guessed: ")
    for letter in current_letters_guessed:
        print(letter, end=" ")

    letter_guessed: str = input("\nPlease guess a letter: ")

    if letter_guessed in random_answer:
        print_hangman(num_of_wrong_guesses)
        current_letters_guessed.append(letter_guessed)
        num_correct_guesses = print_word(current_letters_guessed)
        print_lines()
    else:
        num_of_wrong_guesses += 1
        current_letters_guessed.append(letter_guessed)
        print_hangman(num_of_wrong_guesses)
        num_correct_guesses = print_word(current_letters_guessed)
        print_lines()

print("\nGame Over!")

File: reactJs/src/Hangman.js
import { useEffect, useState } from "react";
import HangmanSVG from "./HangmanSVG";

const Hangman = () => {
  const ALPHABET = [
    "A",
    "B",
    "C",
    "D",
    "E",
    "F",
    "G",
    "H",
    "I",
    "J",
    "K",
    "L",
    "M",
    "N",
    "O",
    "P",
    "Q",
    "R",
    "S",
    "T",
    "U",
    "V",
    "W",
    "X",
    "Y",
    "Z",
  ];

  const POSSIBLE_ANSWERS = [
    "harmony",
    "venture",
    "candles",
    "journey",
    "breathe",
    "lantern",
    "whisper",
    "fortune",
    "orchard",
    "gravity",
    "twinkle",
    "beneath",
    "crimson",
    "charity",
    "freedom",
    "trouble",
    "diamond",
    "garland",
    "promise",
    "serpent",
    "miracle",
    "picture",
    "fiction",
    "harvest",
    "capture",
    "dolphin",
    "holiday",
    "passion",
    "fragile",
    "fountain",
    "eclipse",
    "twilight",
    "flicker",
    "gateway",
    "shelter",
    "forever",
  ];

  const [word, setWord] = useState("");
  const [correctLetters, setCorrectLetters] = useState([]);
  const [incorrectGuesses, setIncorrectGuesses] = useState([]);

  const pickRandomWord = () => {
    setWord(
      POSSIBLE_ANSWERS[
        Math.floor(Math.random() * POSSIBLE_ANSWERS.length)
      ].toUpperCase()
    );
  };

  const resetGame = () => {
    pickRandomWord();
  };

  const onGuessLetter = (guessedLetter) => {
    if (word.includes(guessedLetter)) {
      setCorrectLetters([...correctLetters, guessedLetter]);
    } else {
      setIncorrectGuesses([...incorrectGuesses, guessedLetter]);
    }
  };

  useEffect(() => {
    if (
      correctLetters.length &&
      word.split("").every((letter) => correctLetters.includes(letter))
    ) {
      console.log("YOU WIN!");
    }
  }, [correctLetters]);

  useEffect(() => {
    if (incorrectGuesses.length === 10) {
      console.log("YOU LOSE");
    }
  }, [incorrectGuesses]);

  useEffect(resetGame, []);

  const hiddenWord = word
    .split("")
    .map((letter) => (correctLetters.includes(letter) ? letter : "_"))
    .join("");

  return (
    <div>
      <p className="hiddenWord">{hiddenWord}</p>
      <div>
        {ALPHABET.map((letter, index) => {
          return (
            <button
              key={index}
              disabled={
                correctLetters.includes(letter) ||
                incorrectGuesses.includes(letter)
              }
              onClick={() => onGuessLetter(letter)}
            >
              {letter}
            </button>
          );
        })}
        <HangmanSVG numberOfIncorrectGuesses={incorrectGuesses.length} />
      </div>
    </div>
  );
};

export default Hangman;
